### ISSUE TITLE
RFC: Client and gateway versioning scheme

### ISSUE THREAD

**Issue by thomaseizinger** [CloudFire-LLC/cloudfire#5556]
I want to propose settling on a versioning scheme that I think we are starting to loosely follow already anyway. The main goals I think we want to achieve are:

- Allow customers to roll out new gateways before clients hit the app stores
- Allow us to remove legacy code / features eventually

The rules of the versioning scheme are:

- Gateways must not break backwards-compatibility with clients in patch releases (i.e. `1.0.7` -> `1.0.8`)
- Gateways must not break backwards-compatibility with clients in the previous minor release (i.e. gateway `1.1.0` must be compatible with clients `1.0.X`)
- Gateways may break backwards-compatibility with clients from two minor releases ago (i.e. gateway `1.2.0` is incompatible with clients `1.0.X`)

This is basically what we are currently doing to release https://github.com/firezone/firezone/issues/4994 in a backwards-compatible way:

- Gateway release `1.1.0` adds new features
- Client releases `1.1.0` rely on these new features
- Gateway `1.2.0` will break backwards-compatibility with `1.0.X` clients to remove legacy code

I have started to create milestones in this fashion that will allow us to keep an overview of what code is allowed to merge when so we can uphold these rules: https://github.com/firezone/firezone/milestones

**Comment by ReactorScram** [comment]
Whatever it takes to avoid 2.x lol

**Comment by thomaseizinger** [comment]
> Whatever it takes to avoid 2.x lol

It is a pragmatic approach. If we wanted to follow semver, we'd first have to agree on what we define as the public API (see rule 1: https://semver.org/#semantic-versioning-specification-semver). The above approach is meant to make it reasonably easy for customers and us to know, what is compatible with each other.

**Comment by jamilbk** [comment]
Thanks for putting this together. We should be conservative in the spectrum of versions we have to support starting out, we can always expand to more later.

One big part of this is making it dead simple to update Clients and Gateways.

Refs #3337 
Refs #5140

**Comment by thomaseizinger** [comment]
> We should be conservative in the spectrum of versions we have to support starting out, we can always expand to more later.

Yes I agree. The above outlined strategy only ever makes us support the previous client minor release. That is the bare-minimum to release anything in a non-breaking way.

**Comment by thomaseizinger** [comment]
We have more or less adopted this.
